storm-core/src/jvm/org/apache/storm/ui/filters/AuthorizedUserFilter.java

```diff
@@ -0,0 +1,137 @@
+        NimbusOp annotation = resourceInfo.getResourceMethod().getAnnotation(NimbusOp.class);
+        if (annotation == null) {
+            return;
+        }
+        String op = annotation.value();
+        if (op == null) {
+            return;
+        }
+
+        Map topoConf = null;
+        if (containerRequestContext.getUriInfo().getPathParameters().containsKey("id")) {
+            NimbusClient nimbusClient = NimbusClient.getConfiguredClient(Utils.readStormConfig());
+            try {
+                topoConf = (Map) JSONValue.parse(nimbusClient.getClient().getTopologyConf(
+                        containerRequestContext.getUriInfo().getPathParameters().get("id").get(0)));
+            } catch (TException e) {
+                e.printStackTrace();
+            }
+        }
+
+        ReqContext reqContext = ReqContext.context();
+
+        if (reqContext.isImpersonating()) {
+            if (uiImpersonationHandler != null) {
+                if (!uiImpersonationHandler.permit(reqContext, op, topoConf)) {
+                    Principal realPrincipal = reqContext.realPrincipal();
+                    Principal principal = reqContext.principal();
+                    String user = "unknown";
+                    if (principal != null) {
+                        user = principal.getName();
+                    }
+                    String realUser = "unknown";
+                    if (realPrincipal != null) {
+                        realUser = realPrincipal.getName();
+                    }
+                    InetAddress remoteAddress = reqContext.remoteAddress();
+
+                    throw new IOException(
+                            "user '" + realUser +  "' is not authorized to impersonate user '"
+                            + user + "' from host '" + remoteAddress.toString() + "'. Please"
+                            + "see SECURITY.MD to learn how to configure impersonation ACL."
+                    );
+                }
+
+            LOG.warn(" principal " + reqContext.realPrincipal().toString() + " is trying to impersonate "
+                            + reqContext.principal().toString() + " but "
+                            + conf.get(DaemonConfig.NIMBUS_IMPERSONATION_AUTHORIZER)
+                            + " has no authorizer configured. This is a potential security hole."
+                            + "Please see SECURITY.MD to learn how to configure an impersonation authorizer.");
+            }
+        }
+
+        if (uiAclHandler != null) {
+            if (!uiAclHandler.permit(reqContext, op, null)) {
+                Principal principal = reqContext.principal();
+                String user = "unknown";
+                if (principal != null) {
+                    user = principal.getName();
+                }
+                throw new IOException("UI request '" + op + "' for '" + user + "' user is not authorized");
```

Comment:
Any reason to use IOException instead of AuthorizationException? The meaning of two exceptions are very different.

Reply:
The filter only throws IOExceptions